From f5ba58b45a419011637617de3d05682869de7f73 Mon Sep 17 00:00:00 2001
From: Fang Hui <b31070@freescale.com>
Date: Wed, 4 Dec 2013 15:31:47 +0800
Subject: [PATCH 37/95] ENGR00285701 - 6sl Camera: Recorded video by UVC for
 75s, video stamp drift

mDecodingTimeList should drop stamp when frame is dropped by h264 encoder.

Signed-off-by: Fang Hui <b31070@freescale.com>

Conflicts:
	media/libstagefright/OMXCodec.cpp
	media/libstagefright/codecs/avc/enc/SoftAVCEncoder.cpp
---
 include/media/stagefright/MediaDefs.h                  | 6 ++++++
 media/libstagefright/OMXCodec.cpp                      | 8 ++++++++
 media/libstagefright/codecs/avc/enc/SoftAVCEncoder.cpp | 3 +++
 3 files changed, 17 insertions(+)

diff --git a/include/media/stagefright/MediaDefs.h b/include/media/stagefright/MediaDefs.h
index 21eb04a..bb80594 100644
--- a/include/media/stagefright/MediaDefs.h
+++ b/include/media/stagefright/MediaDefs.h
@@ -14,6 +14,10 @@
  * limitations under the License.
  */
 
+/*
+ * Copyright (C) 2013 Freescale Semiconductor, Inc.
+ */
+
 #ifndef MEDIA_DEFS_H_
 
 #define MEDIA_DEFS_H_
@@ -66,6 +70,8 @@ extern const char *MEDIA_MIMETYPE_TEXT_VTT;
 extern const char *MEDIA_MIMETYPE_TEXT_CEA_608;
 extern const char *MEDIA_MIMETYPE_DATA_TIMED_ID3;
 
+#define OMX_BUFFERFLAG_FRAMEDROP 0xbeafdeed
+
 }  // namespace android
 
 #endif  // MEDIA_DEFS_H_
diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 31ab331..91bcdf3 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -2043,6 +2043,14 @@ void OMXCodec::on_message(const omx_message &msg) {
         {
             IOMX::buffer_id buffer = msg.u.extended_buffer_data.buffer;
 
+			//If the frame is dropped by avcEncoder, need drop stamp in mDecodingTimeList
+            OMX_BUFFERHEADERTYPE *head = (OMX_BUFFERHEADERTYPE *)buffer;
+             if( mIsEncoder && mIsVideo && (!strcasecmp(MEDIA_MIMETYPE_VIDEO_AVC, mMIME)) &&
+                (head->nFlags == OMX_BUFFERFLAG_FRAMEDROP) ) {
+                CODEC_LOGV("drop time stamp %lld us", head->nTimeStamp);
+                getDecodingTimeUs();
+             }
+
             CODEC_LOGV("EMPTY_BUFFER_DONE(buffer: %u)", buffer);
 
             Vector<BufferInfo> *buffers = &mPortBuffers[kPortIndexInput];
diff --git a/media/libstagefright/codecs/avc/enc/SoftAVCEncoder.cpp b/media/libstagefright/codecs/avc/enc/SoftAVCEncoder.cpp
index a9b86da..a79f8df 100644
--- a/media/libstagefright/codecs/avc/enc/SoftAVCEncoder.cpp
+++ b/media/libstagefright/codecs/avc/enc/SoftAVCEncoder.cpp
@@ -623,6 +623,8 @@ void SoftAVCEncoder::onQueueFilled(OMX_U32 /* portIndex */) {
 
                         inQueue.erase(inQueue.begin());
                         inInfo->mOwnedByUs = false;
+
+                        inHeader->nFlags = OMX_BUFFERFLAG_FRAMEDROP;
                         notifyEmptyBufferDone(inHeader);
                         return;
                     }
@@ -671,6 +673,7 @@ void SoftAVCEncoder::onQueueFilled(OMX_U32 /* portIndex */) {
 
         inQueue.erase(inQueue.begin());
         inInfo->mOwnedByUs = false;
+        inHeader->nFlags = 0;
         notifyEmptyBufferDone(inHeader);
 
         outQueue.erase(outQueue.begin());
-- 
1.8.0

